File: articles/cell-signaling-communication-biological-coordination/python/receptor_response.py
"""
Receptor occupancy and Hill response.

Run:
    python python/receptor_response.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
RESPONSE_PATH = ARTICLE_DIR / "data" / "receptor_response.csv"


def main() -> None:
    """Calculate receptor occupancy and cooperative response."""

    observed = pd.read_csv(RESPONSE_PATH)

    L = observed["ligand"].to_numpy(dtype=float)

    Kd = 1.5
    K = 2.0
    n = 3.0

    observed["occupancy"] = L / (Kd + L)
    observed["hill_response"] = L**n / (K**n + L**n)
    observed["residual_observed_minus_hill"] = observed["observed_response"] - observed["hill_response"]

    threshold_row = observed.iloc[(observed["hill_response"] - 0.5).abs().argmin()]

    print(observed.round(4).to_string(index=False))
    print("\nApproximate half-response row:")
    print(threshold_row.round(4).to_string())


if __name__ == "__main__":
    main()

File: articles/cell-signaling-communication-biological-coordination/r/receptor_response.R
# Receptor occupancy and Hill response workflow in R.

receptor_path <- file.path("data", "receptor_response.csv")

if (!file.exists(receptor_path)) {
  receptor_path <- file.path("..", "data", "receptor_response.csv")
}

observed <- read.csv(receptor_path)

Kd <- 1.5
K <- 2.0
n <- 3.0

observed$occupancy <- observed$ligand / (Kd + observed$ligand)
observed$hill_response <- observed$ligand^n / (K^n + observed$ligand^n)
observed$residual_observed_minus_hill <- observed$observed_response - observed$hill_response

print(round(observed, 4))

threshold_row <- observed[which.min(abs(observed$hill_response - 0.5)), ]
print(round(threshold_row, 4))
